File: modules/ble-wrapper/ios/BleWrapperModule.swift
import ExpoModulesCore

public class BleWrapperModule: Module {
    private lazy var nrfManager: NrfBleManager = {
        let manager = NrfBleManager()
        manager.onTemperatureData = { [weak self] temp in
            self?.sendEvent("onTemperatureData", ["temperature": temp])
        }
        manager.onBatteryLevel = { [weak self] level in
            self?.sendEvent("onBatteryLevel", ["level": level])
        }
        manager.onCupStateChanged = { [weak self] state in
            self?.sendEvent("onCupStateChanged", ["state": state])
        }
        manager.onDeviceConnected = { [weak self] in
            self?.sendEvent("onDeviceConnected", ["connected": true])
        }
        manager.onDeviceDisconnected = { [weak self] in
            self?.sendEvent("onDeviceDisconnected", ["connected": false])
        }
        return manager
    }()

    public func definition() -> ModuleDefinition {
        Name("BleWrapper")
        Events(
            "onTemperatureData",
            "onBatteryLevel",
            "onCupStateChanged",
            "onDeviceConnected",
            "onDeviceDisconnected"
        )
        AsyncFunction("connectToXiao") { (promise: Promise) in
            self.nrfManager.connectToXiao(
                onResult: { success in
                    if success { promise.resolve() } else { promise.reject("CONNECT_ERROR", "Connection failed") }
                },
                onFail: { error in promise.reject("CONNECT_ERROR", error) }
            )
        }
        AsyncFunction("disconnect") {
            self.nrfManager.disconnect()
        }
        AsyncFunction("syncTime") { (promise: Promise) in
            self.nrfManager.syncTime(
                onResult: { _ in promise.resolve() },
                onFail: { error in promise.reject("WRITE_ERROR", error) }
            )
        }
        AsyncFunction("setAlarmInterval") { (seconds: Int, promise: Promise) in
            self.nrfManager.setAlarmInterval(seconds: seconds,
                onResult: { _ in promise.resolve() },
                onFail: { error in promise.reject("WRITE_ERROR", error) }
            )
        }
        AsyncFunction("setAlarmMorning") { (hour: Int, second: Int, promise: Promise) in
            self.nrfManager.setAlarmMorning(hour: hour, second: second,
                onResult: { _ in promise.resolve() },
                onFail: { error in promise.reject("WRITE_ERROR", error) }
            )
        }
        AsyncFunction("setAlarmEvening") { (hour: Int, second: Int, promise: Promise) in
            self.nrfManager.setAlarmEvening(hour: hour, second: second,
                onResult: { _ in promise.resolve() },
                onFail: { error in promise.reject("WRITE_ERROR", error) }
            )
        }
        AsyncFunction("readCupState") { (promise: Promise) in
            self.nrfManager.readCupState(
                onResult: { value in promise.resolve(value) },
                onFail: { error in promise.reject("READ_ERROR", error) }
            )
        }
        AsyncFunction("writeCupState") { (state: Int, promise: Promise) in
            self.nrfManager.writeCupState(state: state,
                onResult: { _ in promise.resolve() },
                onFail: { error in promise.reject("WRITE_ERROR", error) }
            )
        }
        AsyncFunction("setAlertsEnabled") { (enabled: Bool, promise: Promise) in
            self.nrfManager.setAlertsEnabled(enabled: enabled,
                onResult: { _ in promise.resolve() },
                onFail: { error in promise.reject("WRITE_ERROR", error) }
            )
        }
        AsyncFunction("findMy") { (promise: Promise) in
            self.nrfManager.findMy(
                onResult: { _ in promise.resolve() },
                onFail: { error in promise.reject("WRITE_ERROR", error) }
            )
        }
        AsyncFunction("readBattery") { (promise: Promise) in
            self.nrfManager.readBattery(
                onResult: { value in promise.resolve(value) },
                onFail: { error in promise.reject("READ_ERROR", error) }
            )
        }
        Function("isConnected") { () -> Bool in
            return self.nrfManager.isConnected
        }
    }
}   

File: modules/ble-wrapper/android/src/main/java/expo/modules/blewrapper/BleWrapperModule.kt
package expo.modules.blewrapper

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise

class BleWrapperModule : Module() {

    private val nrfManager by lazy {
        NrfBleManager(appContext.reactContext!!).also { manager ->
            // Wire up event listeners that push data to JavaScript
            manager.onAccelData = { value ->
                sendEvent("onAccelData", mapOf("value" to value))
            }
            manager.onDeviceConnected = {
                sendEvent("onDeviceConnected", mapOf("connected" to true))
            }
            manager.onDeviceDisconnected = {
                sendEvent("onDeviceDisconnected", mapOf("connected" to false))
            }
        }
    }

    override fun definition() = ModuleDefinition {
        Name("BleWrapper")

        // Events that can be listened to from JS
        Events("onAccelData", "onDeviceConnected", "onDeviceDisconnected")

        // Scan for XIAO_Sense_Accel, returns MAC address
        AsyncFunction("scanForXiao") { promise: Promise ->
            nrfManager.scanForDevice(
                onFound = { address ->
                    promise.resolve(address)
                },
                onScanError = { error ->
                    promise.reject("SCAN_ERROR", error, null)
                }
            )
        }

        // Stop any active BLE scan
        AsyncFunction("stopScan") { promise: Promise ->
            nrfManager.stopScan()
            promise.resolve(null)
        }

        // Connect to device and stay connected (persistent session)
        // Automatically subscribes to accelerometer notifications
        AsyncFunction("connect") { address: String, promise: Promise ->
            nrfManager.connect(
                address = address,
                onResult = { success ->
                    if (success) {
                        promise.resolve(null)
                    } else {
                        promise.reject("CONNECT_ERROR", "Connection failed", null)
                    }
                },
                onFail = { error ->
                    promise.reject("CONNECT_ERROR", error, null)
                }
            )
        }

        // Gracefully disconnect
        AsyncFunction("disconnect") { promise: Promise ->
            nrfManager.disconnect()
            promise.resolve(null)
        }

        // One-shot read of accelerometer (returns "X,Y,Z" string)
        // Device must already be connected via connect()
        AsyncFunction("readAccelerometer") { promise: Promise ->
            nrfManager.readAccelerometer(
                onResult = { value ->
                    promise.resolve(value)
                },
                onFail = { error ->
                    promise.reject("READ_ERROR", error, null)
                }
            )
        }

        // Turn green LED on/off. Device must already be connected.
        AsyncFunction("setLed") { on: Boolean, promise: Promise ->
            nrfManager.writeLed(
                on = on,
                onResult = {
                    promise.resolve(null)
                },
                onFail = { error ->
                    promise.reject("WRITE_ERROR", error, null)
                }
            )
        }

        // Check if currently connected
        Function("isConnected") {
            nrfManager.isConnected
        }
    }
}
